app/models/tag.rb

```diff
@@ -54,8 +53,17 @@ def taggings_count_cache_key
   end
 
   def write_taggings_to_redis(value)
+    # Atomically set the value while extracting the old value.
+    old_redis_value = REDIS_GENERAL.getset("tag_update_#{id}_value", value)
+
+    # If the value hasn't changed from the saved version or the REDIS version,
+    # there's no need to write an update to the database, so let's just bail
+    # out.
+    return value if value == old_redis_value && value == taggings_count_cache
+
+    # If we've reached here, then the value has changed, and we need to make
+    # sure that the new value is written to the database.
     REDIS_GENERAL.sadd("tag_update", id)
```

Comment:
The issue with putting it in `taggings_count=` is that there's not enough information. Suppose that you have a small tag with 10 uses. Now suppose that someone tags a work with that tag, and then immediately removes the tag. (In other words, increase the `taggings_count` by 1, and then immediately decrease the `taggings_count` by 1.) The first change would force it to be added to the `tag_update` set, but wouldn't immediately modify `taggings_count_cache`. During the second update, the code would compute the tag count, discover that the value is 10, cross-check it against the `taggings_count_cache` (which is still 10, because the new value of 11 hasn't been written to MySQL yet), and decide not to write the value to REDIS. The result would be that `taggings_count_cache` would be (incorrectly) set to 11, and the second update wouldn't go through.

The REDIS data isn't authoritative, but if the REDIS value, the `taggings_count_cache` value, and the recomputed value are all the same, then we know that we don't need to change the `taggings_count_cache` anytime soon. (Since the value is correct, and we know that there aren't any impending updates from the values on REDIS that could cause it to be incorrect.)